File: sprint5/By languages/kotlin/lesson10code1.kt
fun heap_add(heap: ArrayList<Int>, key: Int) {
    val index = heap.size + 1
    heap[index] = key
    sift_up(heap, index)
}

fun sift_up(heap: ArrayList<Int>, index: Int) {
    if (index == 1) return

    val parentIndex = index / 2
    if (heap[parentIndex] < heap[index]) {
        val temp = heap[parentIndex]
        heap[parentIndex] = heap[index]
        heap[index] = temp
        sift_up(heap, parentIndex)
    }
}

File: sprint5/By lessons/lesson10code1/lesson10code1.cpp
#include <vector>

void sift_up(std::vector<int>& heap, int index) {
    if (index == 1) return;

    int parentIndex = index / 2;
    if (heap[parentIndex - 1] < heap[index - 1]) {
        std::swap(heap[parentIndex - 1], heap[index - 1]);
        sift_up(heap, parentIndex);
    }
}

void heap_add(std::vector<int>& heap, int key) {
    int index = heap.size() + 1;
    heap.push_back(key);
    sift_up(heap, index);
}
